File: src/main/java/fdu/service/operation/CanProduce.java
package fdu.service.operation;

import fdu.util.ResultCache;
import fdu.util.UserSession;

/**
 * Created by guoli on 2017/5/11.
 */
public interface CanProduce<T> {

    T execute(UserSession user);

    default T executeCached(UserSession user) {
        ResultCache cache = user.getResultCache();
        T result = cache.query(this);
        if (result != null) {
            return result;
        } else {
            result = execute(user);
            cache.commit(this, result);
        }
        return result;
    }

}

File: src/main/scala/fdu/service/operation/operators/CanProduce.scala
package fdu.service.operation.operators

import fdu.exceptions.HiveTableNotFoundException
import fdu.util.UserSession

trait CanProduce[T] {

  @throws(classOf[HiveTableNotFoundException])
  def execute(user: UserSession): T

  @throws(classOf[HiveTableNotFoundException])
  def executeCached(user: UserSession): T = {
    val cache = user.getResultCache
    val original: Option[T] = Option(cache.query(this))
    original match {
      case None =>
        val result = execute(user)
        cache.commit(this, result)
        result
      case Some(res: T) => res
    }
  }
}
